File: Python/two_heaps.py
"""
╔════════════════════════════════════════════════════════╗
║              Two Heaps Pattern (Median)               ║
╠════════════════════════════════════════════════════════╣
║ Description:                                           ║
║ Maintain a max‑heap for the lower half and a min‑heap ║
║ for the upper half of the stream to get median in O(log N) per insert. ║
╠════════════════════════════════════════════════════════╣
║ Flow Diagram (ASCII):                                  ║
║ Stream: 5,15,1,3                                      ║
║                                                      ║
║ Insert 5 → maxH:[5], minH:[] → median=5               ║
║ Insert 15→ maxH:[5], minH:[15] → median=(5+15)/2=10   ║
║ Insert 1 → maxH:[5,1], minH:[15] → median=5           ║
║ Insert 3 → maxH:[3,1], minH:[5,15] → median=(3+5)/2=4 ║
╚════════════════════════════════════════════════════════╝
"""

import heapq

def main():
    maxH = []  # max-heap (invert sign)
    minH = []  # min-heap
    nums = [5, 15, 1, 3]
    for x in nums:
        if not maxH or x < -maxH[0]:
            heapq.heappush(maxH, -x)
        else:
            heapq.heappush(minH, x)
        # Rebalance
        if len(maxH) > len(minH) + 1:
            heapq.heappush(minH, -heapq.heappop(maxH))
        elif len(minH) > len(maxH):
            heapq.heappush(maxH, -heapq.heappop(minH))
        # Median
        if len(maxH) == len(minH):
            print(f"Median: {(-maxH[0] + minH[0])/2.0}")
        else:
            print(f"Median: {-maxH[0]}")

if __name__ == "__main__":
    main() 

File: JAVA/TwoHeaps.java
import java.util.*;

/*
╔════════════════════════════════════════════════════════╗
║              Two Heaps Pattern (Median)               ║
╠════════════════════════════════════════════════════════╣
║ Description:                                           ║
║ Maintain a max‑heap for the lower half and a min‑heap ║
║ for the upper half of the stream to get median in O(log N) per insert. ║
╠════════════════════════════════════════════════════════╣
║ Flow Diagram (ASCII):                                  ║
║ Stream: 5,15,1,3                                      ║
║                                                      ║
║ Insert 5 → maxH:[5], minH:[] → median=5               ║
║ Insert 15→ maxH:[5], minH:[15] → median=(5+15)/2=10   ║
║ Insert 1 → maxH:[5,1], minH:[15] → median=5           ║
║ Insert 3 → maxH:[3,1], minH:[5,15] → median=(3+5)/2=4 ║
╚════════════════════════════════════════════════════════╝
*/

public class TwoHeaps {
    public static void main(String[] args) {
        PriorityQueue<Integer> maxH = new PriorityQueue<>(Collections.reverseOrder()); // lower half
        PriorityQueue<Integer> minH = new PriorityQueue<>(); // upper half
        int[] nums = {5, 15, 1, 3};
        for (int x : nums) {
            // 1) Add new number to one of the heaps
            if (maxH.isEmpty() || x < maxH.peek())
                maxH.add(x);
            else
                minH.add(x);
            // 2) Rebalance to ensure size difference ≤1
            if (maxH.size() > minH.size() + 1) {
                minH.add(maxH.poll());
            } else if (minH.size() > maxH.size()) {
                maxH.add(minH.poll());
            }
            // 3) Compute median
            if (maxH.size() == minH.size())
                System.out.println("Median: " + ((maxH.peek() + minH.peek())/2.0));
            else
                System.out.println("Median: " + maxH.peek());
        }
    }
} 